### PR TITLE
[quant] Implement forward and backward autograd functions for fake quantize

### PR DESCRIPTION
**Pull request by asl3**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81438

### Summary:
This PR implements custom autograd functions for forward and backward to be used in APoT fake quantization. The implementation follows this doc about custom autograd functions: https://pytorch.org/tutorials/beginner/examples_autograd/polynomial_custom_function.html

### Test Plan:
Run tests with: `python test/quantization/core/experimental/test_fake_quantize.py`

### CODE REVIEW

File: test/quantization/core/experimental/test_fake_quantize.py [hunk 1]
```diff
@@ -51,7 +55,7 @@ def test_forward(self):
         X_reduced_precision_fp = apot_fake.forward(torch.clone(X), False)
 
         # get X_expected by converting fp -> apot -> fp to simulate quantize -> dequantize
-        X_to_apot = quantize_APoT(X, alpha, gamma, quantization_levels, level_indices)
+        X_to_apot, mask = quantize_APoT(X, alpha, gamma, quantization_levels, level_indices)
```
Comment (jerryzh168) [inline test/quantization/core/experimental/test_fake_quantize.py:58]: I don't think we should be returning a mask

Comment (asl3) [reply test/quantization/core/experimental/test_fake_quantize.py:58]: @jerryzh168 do you think it makes more sense to just compute mask in the autograd forward method?

File: test/quantization/core/experimental/test_fake_quantize.py [hunk 2]
```diff
@@ -72,5 +76,17 @@ def test_forward_exception(self):
...
             apot_fake.forward(torch.clone(X), False)
 
+    r""" Tests fake quantize helper backward() method
+         using torch.autograd.gradcheck function.
+    """
+    def test_backward(self):
+        input = torch.randn(20, dtype=torch.double, requires_grad=True)
+
+        observer = APoTObserver(b=4, k=2)
+        observer.forward(input)
```
Comment (jerryzh168) [inline test/quantization/core/experimental/test_fake_quantize.py:86]: nit: you can just do `observer(input)`

File: torch/ao/quantization/experimental/fake_quantize_helper.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import torch
+from torch import Tensor
+from torch.ao.quantization.experimental.quantizer import quantize_APoT, dequantize_APoT
+
+class fake_quantize_helper(torch.autograd.Function):
```
Comment (jerryzh168) [inline torch/ao/quantization/experimental/fake_quantize_helper.py:5]: maybe: fake_quantize_function

File: torch/ao/quantization/experimental/quantizer.py [hunk 4]
```diff
@@ -33,10 +33,13 @@ def quantize(self, tensor2quantize: Tensor):
...
-                                                                         self.quantization_levels,
-                                                                         self.level_indices,
-                                                                         self.alpha))
+        tensor2quantize = tensor2quantize.detach().apply_(lambda x: float_to_apot(x,
+                                                                                  self.quantization_levels,
+                                                                                  self.level_indices,
+                                                                                  self.alpha))
+
+        # convert to APoT int representation for dtype
+        tensor2quantize = tensor2quantize.int()
```
Comment (jerryzh168) [inline torch/ao/quantization/experimental/quantizer.py:42]: can we move this to `float_to_apot`?

**Review by jerryzh168 — APPROVED** [review]
looks good overall, had a few comments inline

File: torch/ao/quantization/experimental/quantizer.py (hunk 4 above)
Comment (asl3) [reply torch/ao/quantization/experimental/quantizer.py:42]: float_to_apot deals with each element and i need to change the dtype so i think i need it here

File: torch/ao/quantization/experimental/quantizer.py [hunk 5]
```diff
@@ -33,10 +33,13 @@ def quantize(self, tensor2quantize: Tensor):
         result = torch.tensor([])
 
         # map float_to_apot over tensor2quantize elements
-        tensor2quantize = tensor2quantize.apply_(lambda x: float_to_apot(x,
-                                                                         self.quantization_levels,
-                                                                         self.level_indices,
-                                                                         self.alpha))
+        tensor2quantize = tensor2quantize.detach().apply_(lambda x: float_to_apot(x,
```
Comment (dzdang) [inline torch/ao/quantization/experimental/quantizer.py:36]: have the lambda function issues with jit been resolved?

File: torch/ao/quantization/experimental/fake_quantize_function.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    def forward(ctx,  # type: ignore[override]
+                x: Tensor,
+                alpha: Tensor,
+                gamma: Tensor,
+                quantization_levels: Tensor,
+                level_indices: Tensor) -> Tensor:
+        quantized_result = quantize_APoT(x, alpha, gamma, quantization_levels, level_indices)
+
+        # calculate mask tensor
+        mask = torch.clone(x)
```
Comment (dzdang) [inline torch/ao/quantization/experimental/fake_quantize_function.py:16]: why mask needs to be initialized with a clone of x?

File: torch/ao/quantization/experimental/fake_quantize_helper.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import torch
```
Comment (dzdang) [inline torch/ao/quantization/experimental/fake_quantize_helper.py:1]: what is the purpose of this file? seems to be a duplicate of `fake_quantize_fucntion.py`?

Comment (asl3) [reply torch/ao/quantization/experimental/fake_quantize_helper.py:1]: yep removing this

File: torch/ao/quantization/experimental/fake_quantize_function.py (hunk 6 above)
Comment (asl3) [reply torch/ao/quantization/experimental/fake_quantize_function.py:16]: i used a clone because `apply` modifies tensors in-place. i can just apply on x here though, since the new x would be out of scope after the function ends.

Comment (dzdang) [reply torch/ao/quantization/experimental/fake_quantize_function.py:16]: mask is 0s or 1s. while it's going to get overwritten later, it's not an intuitive initialization. you can just initialize it to 0 or leave it uninitialized

File: torch/ao/quantization/experimental/quantizer.py (hunk 5 above)
Comment (asl3) [reply torch/ao/quantization/experimental/quantizer.py:36]: yep! the jit issues were only caused by the helper function that prints the size of the model, but i can use a different function `state_dict` to get the size of the model

File: torch/ao/quantization/experimental/fake_quantize_function.py (hunk 6 above)
Comment (asl3) [reply torch/ao/quantization/experimental/fake_quantize_function.py]: yep sounds good!
